### PR TITLE
Follow up changes for signal-based inputs

### PR DESCRIPTION
**Pull request by devversion**
See individual commits (language service, docs etc.)

Note that the API docs are still not ideal, but we are incrementally working to support them better. Alternatively we can consider extracting the required function into its own interface for API generation. Lets discuss in follow-ups

### CODE REVIEW

**Review by pkozlowski-opensource — APPROVED** [review]

**Comment by pkozlowski-opensource** [comment]
LGTM but it looks like this PR will need a rebase after recent merges.

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts [hunk 1]
```diff
@@ -626,7 +645,10 @@ export class SymbolBuilder {
     if (ts.isPropertyAccessExpression(node)) {
       tsSymbol = this.getTypeChecker().getSymbolAtLocation(node.name);
     } else if (ts.isElementAccessExpression(node)) {
-      tsSymbol = this.getTypeChecker().getSymbolAtLocation(node.argumentExpression);
+      // Element access expressions need to be resolved with their full context.
+      // Checking the `argumentExpression` will result in the key to be resolved
+      // independently of the surrounding context. I.e. the key itself is simply resolved.
+      tsSymbol = this.getTypeChecker().getSymbolAtLocation(node);
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:651]: This comment is helpful to supprt this change, but I doubt it's useful as a comment for the code without the change context. The whole `else if` case could be dropped instead.

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts [hunk 2]
```diff
@@ -670,3 +692,35 @@ function anyNodeFilter(n: ts.Node): n is ts.Node {
...
+  // 2. Assert that we are dealing with a signal brand symbol access in the argument expression.
+  if (!ts.isElementAccessExpression(expr) ||
+      !ts.isPropertyAccessExpression(expr.argumentExpression)) {
+    return null;
+  }
+
+  // Assert that the property access in the element access is a simple identifier and
+  // refers to `ɵINPUT_SIGNAL_BRAND_WRITE_TYPE`.
+  if (!ts.isIdentifier(expr.argumentExpression.name) ||
+      expr.argumentExpression.name.text !== 'ɵINPUT_SIGNAL_BRAND_WRITE_TYPE') {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:711]: Can this be imported as a symbol and referencing its name? I suppose codegen for the TCB already has a symbol for this, so perhaps that can be used here?

File: packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts [hunk 3]
```diff
@@ -30,13 +30,20 @@ type PropertyDeclarationLike = ts.PropertyDeclaration|ts.AccessorDeclaration;
 /** Type representing either a class declaration ro an interface declaration. */
 type ClassDeclarationLike = ts.ClassDeclaration|ts.InterfaceDeclaration;
 
-/** Type representing either a class member node or an interface member node. */
+/** Type representing either a class or interface member. */
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts:33]: Can this commit be changed to a `refactor` as a non-user-facing change?

**Review by JoostK — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts (hunk 3 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/docs/src/class_extractor.ts:33]: The docs extraction PR (introducing it) was also public-facing, but yeah, let's remove it.

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts (hunk 2 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:711]: Good idea. Using `R3Identifiers`

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts (hunk 1 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts]: Done

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 23eeea5bd750bab2b39cafafd6f91ffec8346aff.

**Pull request closed without merging**